File: daily/2024/962_Maximum_Width_Ramp.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int maxWidthRamp(vector<int>& nums) {
        if (nums.empty()) return 0;
        int ans = 0;
        vector<int> record = {nums[0]};
        vector<int> index = {0};
        for (int i = 1; i < nums.size(); i++){
            auto it = upper_bound(record.begin(), record.end(), nums[i]);
            if (it == record.end()) ans = i;
            else{
                int idx = it - record.begin();
                if (idx == 0){
                    record.insert(it, nums[i]);
                    index.insert(index.begin(), i);
                }
                else{
                    ans = max(ans, i - index[idx-1]);
                }
            }
        }
        return ans;
    }
};

int main(){
    vector<int> nums = {6,0,8,2,1,5};
    // vector<int> nums = {9,8,1,0,1,9,4,0,4,1};
    Solution S;
    cout << S.maxWidthRamp(nums) << endl;
    return 0;
}

File: daily_python/962_Maximum_Width_Ramp.py
import os
import sys
from typing import List

class Solution:
    def maxWidthRamp(self, nums: List[int]) -> int:
        n = len(nums)
        stack = [0]
        ans = 0
        for i in range(1, n):
            if nums[i] < nums[stack[-1]]:
                stack.append(i)
        for i in range(n-1, -1, -1):
            while stack and nums[i] >= nums[stack[-1]]:
                ans = max(ans, i - stack[-1])
                stack.pop()
        return ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nums = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            dataset.append(nums)
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for nums in dataset:
        results.append(solution.maxWidthRamp(nums))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')